java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/protocol/ExtractSuperclassOrInterfaceRefactoring.java

```diff
@@ -0,0 +1,240 @@
+            }
+            if (!members.isEmpty()) {
+                result.add(createCodeAction(Bundle.DN_ExtractInterface(), CodeActionKind.RefactorExtract, null, EXTRACT_INTERFACE_REFACTORING_COMMAND, uri, elementItem, members));
+            }
+        }
+        return result;
+    }
+
+    @Override
+    public Set<String> getCommands() {
+        return commands;
+    }
+
+    @Override
+    @NbBundle.Messages({
+        "DN_SelectMembersToExtract=Select members to extract",
+        "DN_SelectClassName=Select class name",
+        "DN_SelectInterfaceName=Select interface name",
+    })
+    public CompletableFuture<Object> processCommand(NbCodeLanguageClient client, String command, List<Object> arguments) {
+        if (arguments.size() > 2) {
+            String uri = gson.fromJson(gson.toJson(arguments.get(0)), String.class);
+            QuickPickItem type = gson.fromJson(gson.toJson(arguments.get(1)), QuickPickItem.class);
+            List<QuickPickItem> members = Arrays.asList(gson.fromJson(gson.toJson(arguments.get(2)), QuickPickItem[].class));
+            client.showQuickPick(new ShowQuickPickParams(Bundle.DN_SelectMembersToExtract(), true, members)).thenAccept(selected -> {
+                if (selected != null && !selected.isEmpty()) {
+                    String label = EXTRACT_SUPERCLASS_REFACTORING_COMMAND.equals(command) ? Bundle.DN_SelectClassName() : Bundle.DN_SelectInterfaceName();
+                    String value = EXTRACT_SUPERCLASS_REFACTORING_COMMAND.equals(command) ? "NewClass" : "NewInterface";
+                    client.showInputBox(new ShowInputBoxParams(label, value)).thenAccept(name -> {
+                        if (name != null && !name.isEmpty()) {
+                            extract(client, uri, command, type, selected, name);
+                        }
+                    });
+                }
+            });
+        } else {
+            client.showMessage(new MessageParams(MessageType.Error, String.format("Illegal number of arguments received for command: %s", command)));
+        }
+        return CompletableFuture.completedFuture(true);
+    }
+
+    private void extract(NbCodeLanguageClient client, String uri, String command, QuickPickItem source, List<QuickPickItem> members, String name) {
+        try {
+            FileObject file = Utils.fromUri(uri);
+            ClasspathInfo info = ClasspathInfo.create(file);
+            ElementHandle handle = gson.fromJson(gson.toJson(source.getUserData()), ElementData.class).toHandle();
+            AbstractRefactoring refactoring;
+            if (EXTRACT_SUPERCLASS_REFACTORING_COMMAND.equals(command)) {
+                List<MemberInfo<ElementHandle<Element>>> memberHandles = new ArrayList<>();
+                JavaSource js = JavaSource.forFileObject(file);
+                if (js == null) {
+                    throw new IOException("Cannot get JavaSource for: " + uri);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Exception [IOException] is explicitly thrown on this line but is caught in the outer catch clause with generic exception on line [236].  Consider adding a catch clause of this specific exception in the outer try statement.